### ISSUE TITLE
SPARQL: Standardize free text search of RDF data

### ISSUE THREAD

**Issue by dbooth-boston** [w3c/sparql-12#40]
Several RDF stores support free text search, but there's no standard way to do it.
Proposed by Kjetil Kjernsmo in W3C Graph workshop lightning talk: https://www.w3.org/Data/events/data-ws-2019/assets/lightning/KjetilKjernsmo.pptx

**Comment by seralf** [comment]
Hi one of the approach for handling this feature could be thinking about a standardized way to use external indexing services, by means of a specific usage of the `SERVICE` statement (#10). 

For example for blazegraph:
+ https://wiki.blazegraph.com/wiki/index.php/FullTextSearch
+ https://wiki.blazegraph.com/wiki/index.php/ExternalFullTextSearch
+ https://wiki.blazegraph.com/wiki/index.php/SOLR_External_Fulltext_Search

this document seems to have a nice overview of various different implementations:
https://www.ida.liu.se/research/semanticweb/events/Material/ISWC2017-SemDataMgmtTutorial-Part4-Searching.pdf

**Comment by afs** [comment]
Removing "SPARQL: " on transferred issue.

**Comment by kjetilk** [comment]
Indeed, this was hard to do in SPARQL 1.1, since it would require some survey work to understand the specific use cases. 

However, I think the key is that nearly every web site and app has a search bar. I've seen a lot of architectural complexity added just to support that on backends. Some basic support seems to me to be very important, but the scope should be limited.

**Comment by seralf** [comment]
Another approach to handle FTS is used in [Halyard](https://merck.github.io/Halyard/) project:
```SPARQL
SELECT ?subj ?pred
WHERE {
    ?subj ?pred "(search~1 algorithm~1) AND (grant ingersoll)"^^halyard:search
}
```
https://merck.github.io/Halyard/usage.html#cooperation-with-elasticsearch

in this case the actual FTS is done on literals externally (on an external Elasticsearch instance), similarly to blazegraph. The syntax seems rather different from the other ones cited before.

**Comment by afs** [comment]
Full Text Search is important. Let's find what to consolidate around as a new feature and what to leave flexible. While perfect functionality alignment isn't possible if system use external libraries but is there one "text search" feature which has some common syntax across systems?

We have two issues for better calling external functionality in "multiple returns from a function #6" and 
"named argument #48".

Are they enough as a mechanism?

Or maybe FTS is so important it deserves special syntax , with URI to a specific FTS implementation (even if that syntax is calling the same mechanisms) or would different systems be giving a different URI to their capabilities and the user has to see this?

Rough example: `TEXT (?literal, ?score) OF ( "search terms";  "limit"=10 ; score="50")`

**Comment by jpcs** [comment]
The XQuery Fulltext extension is at it's core not specific to XML, and could easily be lifted wholesale into SPARQL. It took a lot of time and effort to develop - and it would be good not to duplicate this effort.

https://www.w3.org/TR/xpath-full-text-10/

**Comment by namedgraph** [comment]
My attempt to list triplestores with full-text search functionality:
https://github.com/AtomGraph/LinkedDataHub/wiki/Full-text-search-support-in-triplestores

**Comment by VladimirAlexiev** [comment]
@namedgraph The entry for GDB links to deprecated old functionality. Please add these links:
https://graphdb.ontotext.com/documentation/enterprise/general-full-text-search-with-connectors.html
https://graphdb.ontotext.com/documentation/enterprise/lucene-graphdb-connector.html
https://graphdb.ontotext.com/documentation/enterprise/solr-graphdb-connector.html
https://graphdb.ontotext.com/documentation/enterprise/elasticsearch-graphdb-connector.html
http://platform.ontotext.com/search/index.html
